src/geometry.js

```diff
@@ -374,33 +374,67 @@ var g = (function() {
+                    this.start.y === l.start.y &&
+                    this.start.x === l.start.x &&
+                    this.end.y === l.end.y;
+        },
+
+        // @return {point} Point where I'm intersecting a line.
+        // @return [point] Points where I'm intersecting a rectangle.
         // @see Squeak Smalltalk, LineSegment>>intersectionWith:
-        intersection: function(l) {
-            var pt1Dir = Point(this.end.x - this.start.x, this.end.y - this.start.y);
-            var pt2Dir = Point(l.end.x - l.start.x, l.end.y - l.start.y);
-            var det = (pt1Dir.x * pt2Dir.y) - (pt1Dir.y * pt2Dir.x);
-            var deltaPt = Point(l.start.x - this.start.x, l.start.y - this.start.y);
-            var alpha = (deltaPt.x * pt2Dir.y) - (deltaPt.y * pt2Dir.x);
-            var beta = (deltaPt.x * pt1Dir.y) - (deltaPt.y * pt1Dir.x);
-
-            if (det === 0 ||
-                alpha * det < 0 ||
-                beta * det < 0) {
-                // No intersection found.
-                return null;
-            }
-            if (det > 0) {
-                if (alpha > det || beta > det) {
+        intersect: function(l) {
+
+            if (l instanceof Line) {
+                // Passed in parameter is a line.
+
+                var pt1Dir = Point(this.end.x - this.start.x, this.end.y - this.start.y);
+                var pt2Dir = Point(l.end.x - l.start.x, l.end.y - l.start.y);
+                var det = (pt1Dir.x * pt2Dir.y) - (pt1Dir.y * pt2Dir.x);
+                var deltaPt = Point(l.start.x - this.start.x, l.start.y - this.start.y);
+                var alpha = (deltaPt.x * pt2Dir.y) - (deltaPt.y * pt2Dir.x);
+                var beta = (deltaPt.x * pt1Dir.y) - (deltaPt.y * pt1Dir.x);
+
+                if (det === 0 ||
+                    alpha * det < 0 ||
+                    beta * det < 0) {
+                    // No intersection found.
                     return null;
                 }
-            } else {
-                if (alpha < det || beta < det) {
-                    return null;
+                if (det > 0) {
+                    if (alpha > det || beta > det) {
+                        return null;
+                    }
+                } else {
+                    if (alpha < det || beta < det) {
+                        return null;
+                    }
                 }
+                return Point(this.start.x + (alpha * pt1Dir.x / det),
+                             this.start.y + (alpha * pt1Dir.y / det));
+
+            } else if (l instanceof Rect) {
+                // Passed in parameter is a rectangle.
+
+                var r = l;
+                var rectLines = [ r.topLine(), r.rightLine(), r.bottomLine(), r.leftLine() ];
+                var points = [];
+                var pt, i;
+
+                for (i = 0; i < rectLines.length; i ++) {
+                    pt = this.intersect(rectLines[i]);
+                    if (pt !== null && points.indexOf(pt) < 0) {
```

Comment:
Fixed it in https://github.com/clientIO/joint/pull/497/commits/a5f2bda1fb24d123e923edc8fbf998830f6b9334. Will look at the line intersection logic next.